Miner Actor events for Sector lifecycle

aarshkshah1992 opened the pull request:
For https://github.com/filecoin-project/builtin-actors/issues/1464.

This PR implements the following sector lifecycle events for the Miner Actor:

1. Sector Precommitted
2. Sector Activated
3. Sector Updated
4. Sector Temrinated

aarshkshah1992 commented on actors/miner/src/lib.rs:
```diff
@@ -4935,7 +4973,17 @@ fn activate_new_sector_infos(
     data_activations: Vec<DataActivationOutput>,
     pledge_inputs: &NetworkPledgeInputs,
     info: &MinerInfo,
+    emit_event: bool,
```
@anorth We need this as we only want the emit here if this call is from the cron callback to `CONFIRM_SECTOR_PROOFS_VALID` which in turn is scheduled by the vanilla prove commit method.

For the prove commit aggregate method, we don't want to emit the event here as we also need to check if we have enough funds for gas fees. Doing it this way makes it work nicely with the tests.

anorth replied:
I don't think this is the right trade-off. Emitting the events here is a good place to ensure that future code changes will still emit events exactly according to activations, and will simplify code elsewhere. There are more paths into sector activation coming in `integration/direct-onboarding`.

We don't need to make checks about funds etc before emitting events - if those checks fail the whole thing will abort anyway. We may need to update testing support code to allow us to write the main code in the straightforward way.

aarshkshah1992 replied:
This is done. We no longer have this flag.

anorth commented on actors/miner/src/emit.rs:
```diff
@@ -0,0 +1,27 @@
+use fvm_shared::sector::SectorID;
+
+/// Indicates a sector has been pre-committed.
+pub fn sector_precommited(rt: &impl Runtime, id: SectorID) -> Result<(), ActorError> {
```
I know the SectorID type has exactly what you want, but please pass two separate parameters here for the miner actor ID and sector number separately. The SectorID type only exists as part of SealVerifyInfo; it's defined outside of our repo and is unused except when forced by the FVM SDK API. I was genuinely surprised to see it as I'd forgotten it exists.

aarshkshah1992 replied:
This is done.

anorth commented on actors/miner/src/emit.rs:
```diff
@@ -0,0 +1,27 @@
+}
+
+/// Indicates a sector has been proven.
+pub fn sector_proven(rt: &impl Runtime, id: SectorID) -> Result<(), ActorError> {
```
Please rename to sector-activated. Today these always happen at the same time, but in the future we may separate them. If we do, it's likely that most users of the event really care about the activation, rather than an un-activated proof. We'll reserve sector-proven for that case.

aarshkshah1992 replied:
Done.

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -852,12 +854,16 @@ impl Actor {
             activate_deals(rt, &data_activations, compute_commd)?;
         let successful_activations = batch_return.successes(&precommits_to_confirm);
 
+        let proven_sector_nums =
```
activated_sector_nums

aarshkshah1992 replied:
Done.

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -878,6 +884,11 @@ impl Actor {
+
+        for sector in proven_sector_nums {
+            emit::sector_proven(rt, SectorID { miner: miner_actor_id, number: sector })?;
+        }
```
Nit:I suggest moving this up right under activate_new_sector_infos

aarshkshah1992 replied:
Done.

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -1484,6 +1495,16 @@ impl Actor {
         rt: &impl Runtime,
         sectors: Vec<SectorPreCommitInfoInner>,
     ) -> Result<(), ActorError> {
+        let miner_actor_id: u64 = if let Payload::ID(i) = rt.message().receiver().payload() {
```
This is guaranteed to be an ID address, so just unwrap it.

aarshkshah1992 replied:
Done.

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -1635,6 +1656,8 @@ impl Actor {
             let sector_weight_for_deposit = qa_power_max(info.sector_size);
             let deposit_req = pre_commit_deposit_for_power(&reward_stats.this_epoch_reward_smoothed, &power_total.quality_adj_power_smoothed, &sector_weight_for_deposit);
 
+            let mut precommited_sectors = vec![];
```
We already have a sector_numbers variable. It's a bitfield, so slightly inefficient, but still easily iterable. So this extra collection is unnecessary.

aarshkshah1992 replied:
Done.

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -1720,9 +1745,20 @@ impl Actor {
             needs_cron = !state.deadline_cron_active;
             state.deadline_cron_active = true;
+
+            for sector_num in precommited_sectors.into_iter() {
+                let sector_id = SectorID{
+                    miner: miner_actor_id,
+                    number: sector_num,
+                };
+                emit::sector_precommited(rt, sector_id)?;
+            }
```
Move this up a few lines so it's directly after the state updates.

aarshkshah1992 replied:
Done.

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -4935,7 +4973,17 @@ fn activate_new_sector_infos(
     info: &MinerInfo,
+    emit_event: bool,
 ) -> Result<(), ActorError> {
+    let miner_actor_id: u64 = if let Payload::ID(i) = rt.message().receiver().payload() {
```
Just unwrap

anorth commented on actors/miner/tests/compact_sector_numbers_tests.rs:
```diff
@@ -49,7 +49,7 @@ mod compact_sector_numbers_test {
             expect_abort(
                 ExitCode::USR_ILLEGAL_ARGUMENT,
-                h.pre_commit_sector(&rt, precommit, util::PreCommitConfig::default(), false),
+                h.pre_commit_sector(&rt, precommit, util::PreCommitConfig::default(), false, false),
```
We shouldn't have to pass this flag. This pattern doesn't scale to all the possible expectations and abort points that a method might hit.

The mock should always set expectations for the events. If the method aborts before emitting them, then at line 54 you should `rt.reset()` to void any remaining expectations. Most places will probably already do this, but this one forgot.

aarshkshah1992 replied:
This has been fixed in all tests. Done.

anorth commented on integration_tests/src/tests/commit_post_test.rs:
```diff
@@ -720,7 +720,7 @@ pub fn aggregate_one_precommit_expires_test(v: &dyn VM) {
         v,
         &worker,
-        &robust_addr,
+        &miner_addr,
```
Why was this change necessary? It shouldn't be

aarshkshah1992 replied:
@anorth This is because the Test Runtime always expects the receiver address to be an ID Address.


https://github.com/filecoin-project/builtin-actors/blob/7500bab29e5b14ed433f19145f27c53034c6bead/test_vm/src/messaging.rs#L643

aarshkshah1992 replied:
Can we change the emitter field here to be just an `Address` instead ? We can then worry about what to do with it once we get to the integration tests.

anorth replied:
The event emitter must always be an ID. The runtime should always give the actor an ID address as `receiver()`. The problem is somewhere in our TestVM which should resolve an address to an ID address before making it available as the receiver (but not too soon, in case the actor doesn't yet exist and needs to be created on the fly to receive the message). It's important that we can support calls to non-ID addresses like this but then have everything still work.

anorth reviewed (changes requested):
Please change the testing patterns to remove the boolean configuration of which expectations to set, and instead reset the mock after a method aborts.

We also need events for data activation via ProveReplicaUpdates. I can't tell from the PR description whether you intended this to be the only PR to miner actor, or would follow-up. But I think it probably makes sense to do it all at once here so we can confirm the same patterns work for both.

aarshkshah1992 commented on actors/miner/src/emit.rs:
```diff
@@ -0,0 +1,27 @@
+use fvm_shared::sector::SectorID;
+
+/// Indicates a sector has been pre-committed.
+pub fn sector_precommited(rt: &impl Runtime, id: SectorID) -> Result<(), ActorError> {
```
Done.

aarshkshah1992 commented on actors/miner/src/emit.rs:
```diff
@@ -0,0 +1,27 @@
+pub fn sector_precommited(rt: &impl Runtime, id: SectorID) -> Result<(), ActorError> {
+    rt.emit_event(
+        &EventBuilder::new()
+            .typ("sector-precommited")
```
Done.

aarshkshah1992 commented:
@anorth 

I have addressed your review and have added events for the **_Sector Updated_** and **_Sector Terminated_** transitions. Some notes/questions:

1. The failing integration tests are because of https://github.com/filecoin-project/builtin-actors/pull/1462#discussion_r1372230713. Should we change the test runtime to accept non-ID addresses when events are emitted or should we change the emitter type there to an `Address` instead of `ActorId` ? I am not sure of the tradeoffs here.

2. There are no unit tests for the **_ProveReplicaUpdate*_** method and so no event emission tests for that. I think we will have to cover this when we get to integration tests as we do have integration tests for this work. Any suggestions/concerns ?

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -878,6 +880,7 @@ impl Actor {
         }
         burn_funds(rt, aggregate_fee)?;
         state.check_balance_invariants(&rt.current_balance()).map_err(balance_invariants_broken)?;
+
```
Nit: unnecessary blank line. Some others similar below.

aarshkshah1992 replied:
This is done.

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -1484,6 +1495,8 @@ impl Actor {
         sectors: Vec<SectorPreCommitInfoInner>,
     ) -> Result<(), ActorError> {
+        let miner_actor_id: u64 = rt.message().receiver().id().unwrap();
+
```
Nit: unnecessary blank

aarshkshah1992 replied:
Fixed.

anorth commented on actors/miner/tests/prove_commit.rs:
```diff
@@ -511,11 +511,18 @@ fn drop_invalid_prove_commit_while_processing_valid_one() {
     assert!(st.get_sector(&rt.store, sector_no_a).unwrap().is_none());
     assert!(st.get_sector(&rt.store, sector_no_b).unwrap().is_some());
     h.check_state(&rt);
+    rt.reset();
```
This call didn't abort, so this should `rt.verify()` instead. Or you can skip it because it's automatic when dropped, which happens immediately next. If you added this because expectations are failing, something is wrong.

aarshkshah1992 replied:
Fixed, we don't need this. It's a remanent of something I was debugging earlier.

anorth commented on actors/miner/tests/util.rs:
```diff
@@ -620,6 +620,16 @@ impl ActorHarness {
+                    .typ("sector-precommitted")
+                    .field_indexed("miner", &RECEIVER_ID)
+                    .field_indexed("sector", &si.sector_number)
+                    .build()?,
```
There's enough repeated code here that it's probably just worth adding a method to factor it out, like `expect_event(rt, "sector-precommitted", &RECEIVER_ID, &si.sector_number)`. This can be broadly useful because all the miner events have the same schema.

aarshkshah1992 replied:
Refactored this into it's function.

anorth reviewed (commented):
I want us to figure out the address resolution in tests, but otherwise this looks good.

aarshkshah1992 commented:
@anorth I've fixed the problem of the receiver address not being an ID address in the TestVM. Things work well now without any changes to the integration tests and so have rolled those changes back.
